reconcile/utils/slack_api.py

```diff
@@ -16,36 +20,84 @@ class UsergroupNotFoundException(Exception):
     pass
 
 
-class SlackAPICallException(Exception):
-    """Raised for general error cases when calling the Slack API."""
-
-
-class SlackAPIRateLimitedException(SlackAPICallException):
-    """Raised when a call to the Slack API has been rate-limited."""
+class ServerErrorRetryHandler(RetryHandler):
+    """Retry handler for 5xx errors."""
+    def _can_retry(
+        self,
+        *,
+        state: RetryState,
+        request: HttpRequest,
+        response: Optional[HttpResponse] = None,
+        error: Optional[Exception] = None
+    ) -> bool:
+        return response is not None and response.status_code >= 500
 
 
 class SlackApi:
     """Wrapper around Slack API calls"""
 
-    def __init__(self, workspace_name, token,
-                 settings=None,
+    def __init__(self,
+                 workspace_name: str,
+                 token: Mapping[str, str],
+                 settings: Optional[Mapping[str, Any]] = None,
                  init_usergroups=True,
-                 **chat_kwargs):
+                 channel: Optional[str] = None,
+                 **chat_kwargs) -> None:
+        """
+        :param workspace_name: Slack workspace name (ex. coreos)
+        :param token: data to pass to SecretReader.read() to get the token
+        :param settings: settings to pass to SecretReader
+        :param init_usergroups: whether or not to get a list of all Slack
+        usergroups when instantiated
+        :param channel: the Slack channel to post messages to
+        :param chat_kwargs: any other kwargs that can be used to post Slack
+        channel messages
+        """
         self.workspace_name = workspace_name
+
         secret_reader = SecretReader(settings=settings)
```

Comment:
This one tears me apart. Is this the right moment to move all side effects from the constructor and use this API in a context manager?

Reply:
What are you thinking with respect to the context manager?

I think the issue of `SecretReader` and more generally creating resources in `__init__` is more widespread and might be best handled in a more general way.  It isn't clear to me that it's much better to have `SecretReader` created by `__init__` indirectly (by calling another method) than directly.

Instead of creating 100 SecretReaders, one for each class, and dealing with mocks everywhere, we could handle this in a settings module or some equivalent that moves in the direction of dependency injection.  A dummy example:

```python
"""settings.py"""

class Settings:

@property
def slack_token() -> str:
    # Ignoring what __init__ of this would look like for now, assume there
    # is a SecretReader created somehow
    self._secret_reader.read(SLACK_SECRET_LOCATION)

settings = Settings()



"""slack_usergroups.py (or anywhere else SlackApi is used)"""

from reconcile.settings import settings

slack_api = SlackApi('workspace', settings.slack_token)
```

The SlackApi class no longer needs "settings" directly passed into it, nor does it need a `SecretReader` object at all.  I think this would really lead to cleaner and more testable code.  I'm willing to put together a quick PR in the next few days to show how we could do this to clean up the `__init__` in a few classes.  It won't solve every issues in the constructors, but it could be a move in the right direction.

Reply:
Spoke to @Piojo about this offline.  He might also respond, but he agreed that passing the token in directly would be better.  I'll keep this as a separate task for now to avoid delaying some of the benefits that we'll get from this upgrade.